File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Playback/PlaybackViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for audio playback using SonixPlayer.
///
/// ## VoxaTrace Integration (~20 lines)
/// ```swift
/// // 1. Create player with Config + Factory pattern
/// let config = SonixPlayerConfig.Builder()
///     .volume(volume).pitch(pitch).loopCount(loopCount)
///     .onComplete { }.onError { }.build()
/// player = try await SonixPlayer.create(source: path, config: config, audioSession: .playback)
///
/// // 2. Observe state
/// player.observeIsPlaying { }
/// player.observeCurrentTime { }
///
/// // 3. Control playback
/// player.play() / player.pause() / player.stop() / player.seek(positionMs:)
/// ```
@MainActor
final class PlaybackViewModel: ObservableObject {

    // MARK: - Published State

    @Published private(set) var isPlaying = false
    @Published private(set) var currentTimeMs: Int64 = 0
    @Published private(set) var durationMs: Int64 = 0
    @Published private(set) var isLoaded = false
    @Published private(set) var status = "Ready"
    @Published var volume: Float = 0.8
    @Published var pitch: Float = 0
    @Published var loopCount: Int32 = 1

    // MARK: - Computed Properties

    var seekProgress: Double {
        durationMs > 0 ? Double(currentTimeMs) / Double(durationMs) : 0
    }

    var formattedCurrentTime: String { formatTime(currentTimeMs) }
    var formattedDuration: String { formatTime(durationMs) }

    // MARK: - Private

    private var player: SonixPlayer?
    private var playingTask: Task<Void, Never>?
    private var timeTask: Task<Void, Never>?

    // MARK: - Init

    init() {
        // Defer async initialization using DispatchQueue (Task @MainActor doesn't update UI properly)
        DispatchQueue.main.async { [weak self] in
            self?.loadPlayer()
        }
    }

    private func loadPlayer() {
        status = "Loading..."

        Task { [weak self] in
            guard let self = self else { return }
            await self.doInitializePlayer()
        }
    }

    private func doInitializePlayer() async {
        guard let assetPath = copyAssetToFile(name: "sample", ext: "m4a") else {
            Log.e(.playback, "Asset 'sample.m4a' not found in bundle")
            DispatchQueue.main.async { [weak self] in
                self?.status = "Asset not found"
            }
            return
        }

        do {
            let config = SonixPlayerConfig.Builder()
                .volume(volume)
                .pitch(pitch)
                .loopCount(loopCount)
                .onComplete { [weak self] in
                    Log.i(.playback, "Playback completed")
                    Task { @MainActor in
                        self?.status = "Playback complete"
                    }
                }
                .onLoopComplete { loopIndex, totalLoops in
                    Log.i(.playback, "Loop \(loopIndex + 1)/\(totalLoops) completed")
                }
                .onError { error in
                    Log.e(.playback, "Playback error: \(error)")
                }
                .build()

            let newPlayer = try await SonixPlayer.create(source: assetPath, config: config, audioSession: .playback)
            Log.i(.playback, "Player initialized: duration=\(newPlayer.duration)ms")

            DispatchQueue.main.async { [weak self] in
                self?.player = newPlayer
                self?.durationMs = newPlayer.duration
                self?.isLoaded = true
                self?.status = "Loaded: sample.m4a"
                self?.observePlaybackState(newPlayer)
            }
        } catch {
            Log.e(.playback, "Failed to create player", error: error)
            DispatchQueue.main.async { [weak self] in
                self?.status = "Error: \(error.localizedDescription)"
            }
        }
    }

    // MARK: - Lifecycle

    func onAppear() {
        // Initialization now happens in init
    }

    func onDisappear() {
        cancelObservers()
        player?.release()
        player = nil
    }

    // MARK: - Actions

    func play() {
        player?.play()
    }

    func pause() {
        player?.pause()
    }

    func stop() {
        player?.stop()
        currentTimeMs = 0
    }

    func seek(to progress: Double) {
        let seekPos = Int64(progress * Double(durationMs))
        player?.seek(positionMs: seekPos)
    }

    func setVolume(_ newVolume: Float) {
        volume = newVolume
        player?.volume = newVolume
    }

    func setPitch(_ newPitch: Float) {
        pitch = newPitch
        player?.pitch = newPitch
    }

    func setLoopCount(_ count: Int32) {
        loopCount = count
        player?.loopCount = count
    }

    func fadeIn() {
        Task {
            try? await player?.fadeIn(targetVolume: 1.0, durationMs: 1000)
        }
    }

    func fadeOut() {
        Task {
            try? await player?.fadeOut(durationMs: 1000)
        }
    }

    // MARK: - Private Methods

    private func observePlaybackState(_ player: SonixPlayer) {
        playingTask = player.observeIsPlaying { [weak self] playing in
            self?.isPlaying = playing
        }
        timeTask = player.observeCurrentTime { [weak self] time in
            self?.currentTimeMs = time
        }
    }

    private func cancelObservers() {
        playingTask?.cancel()
        timeTask?.cancel()
    }

    private func formatTime(_ ms: Int64) -> String {
        let seconds = (ms / 1000) % 60
        let minutes = (ms / 1000) / 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/playback/viewmodel/PlaybackViewModel.kt
package com.musicmuni.voxatrace.demo.sections.playback.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.sonix.AudioMode
import com.musicmuni.voxatrace.sonix.SonixPlayer
import com.musicmuni.voxatrace.sonix.SonixPlayerConfig
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * ViewModel for audio playback using SonixPlayer.
 *
 * ## VoxaTrace Integration (~20 lines)
 * ```kotlin
 * // 1. Create player with Config + Factory pattern
 * val config = SonixPlayerConfig.Builder()
 *     .volume(0.8f)
 *     .pitch(-2f)
 *     .loopCount(3)
 *     .onComplete { }
 *     .onLoopComplete { loopIndex, totalLoops -> }
 *     .onError { }
 *     .build()
 * val player = SonixPlayer.create(source, config, audioSession = AudioMode.PLAYBACK)
 *
 * // 2. Observe state via StateFlows
 * player.isPlaying.collect { }
 * player.currentTime.collect { }
 *
 * // 3. Control playback
 * player.play() / player.pause() / player.stop() / player.seek(ms)
 * player.volume = 0.5f / player.pitch = 2f
 * ```
 */
class PlaybackViewModel : ViewModel() {

    // MARK: - Published State

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _currentTimeMs = MutableStateFlow(0L)
    val currentTimeMs: StateFlow<Long> = _currentTimeMs.asStateFlow()

    private val _durationMs = MutableStateFlow(0L)
    val durationMs: StateFlow<Long> = _durationMs.asStateFlow()

    private val _volume = MutableStateFlow(0.8f)
    val volume: StateFlow<Float> = _volume.asStateFlow()

    private val _pitch = MutableStateFlow(0f)
    val pitch: StateFlow<Float> = _pitch.asStateFlow()

    private val _loopCount = MutableStateFlow(1)
    val loopCount: StateFlow<Int> = _loopCount.asStateFlow()

    private val _status = MutableStateFlow("Ready")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _isLoaded = MutableStateFlow(false)
    val isLoaded: StateFlow<Boolean> = _isLoaded.asStateFlow()

    // MARK: - Private State

    private var player: SonixPlayer? = null
    private val observerJobs = mutableListOf<Job>()

    // MARK: - Actions

    fun initialize(context: Context) {
        if (_isLoaded.value) return

        viewModelScope.launch {
            try {
                _status.value = "Loading..."

                // Copy asset to file on IO dispatcher
                val assetFile = withContext(Dispatchers.IO) {
                    copyAssetToFile(context, "sample.m4a")
                }

                // Create player using Config + Factory pattern (ADR-001)
                val playerConfig = SonixPlayerConfig.Builder()
                    .volume(_volume.value)
                    .pitch(_pitch.value)
                    .loopCount(_loopCount.value)
                    .onComplete {
                        Napier.d("Playback completed!")
                        _status.value = "Playback complete"
                    }
                    .onLoopComplete { loopIndex, totalLoops ->
                        Napier.d("Loop $loopIndex of $totalLoops completed")
                    }
                    .onError { error ->
                        Napier.e("Playback error: $error")
                        _status.value = "Error: $error"
                    }
                    .build()

                val newPlayer = SonixPlayer.create(
                    source = assetFile.absolutePath,
                    config = playerConfig,
                    audioSession = AudioMode.PLAYBACK
                )

                player = newPlayer
                _durationMs.value = newPlayer.duration
                _isLoaded.value = true
                _status.value = "Loaded: sample.m4a"
                Napier.d("Audio loaded, duration: ${newPlayer.duration} ms")

                setupObservers(newPlayer)
            } catch (e: Exception) {
                Napier.e("Player init failed", e)
                _status.value = "Error: ${e.message}"
            }
        }
    }

    fun play() {
        player?.play()
    }

    fun pause() {
        player?.pause()
    }

    fun stop() {
        player?.stop()
        _currentTimeMs.value = 0
    }

    fun seek(fraction: Float) {
        val seekPos = (fraction * _durationMs.value).toLong()
        player?.seek(seekPos)
    }

    fun setVolume(newVolume: Float) {
        _volume.value = newVolume
        player?.let { it.volume = newVolume }
    }

    fun setPitch(newPitch: Float) {
        _pitch.value = newPitch
        player?.let { it.pitch = newPitch }
    }

    fun setLoopCount(count: Int) {
        _loopCount.value = count
        player?.let { it.loopCount = count }
    }

    fun fadeIn() {
        viewModelScope.launch {
            player?.fadeIn(1.0f, 1000)
        }
    }

    fun fadeOut() {
        viewModelScope.launch {
            player?.fadeOut(1000)
        }
    }

    // MARK: - Private Methods

    private fun setupObservers(player: SonixPlayer) {
        observerJobs.add(viewModelScope.launch {
            player.isPlaying.collect { playing ->
                _isPlaying.value = playing
            }
        })

        observerJobs.add(viewModelScope.launch {
            player.currentTime.collect { time ->
                _currentTimeMs.value = time
            }
        })
    }

    override fun onCleared() {
        super.onCleared()
        observerJobs.forEach { it.cancel() }
        player?.release()
    }

    // MARK: - Helpers

    private fun copyAssetToFile(context: Context, assetName: String): File {
        val file = File(context.cacheDir, assetName)
        if (!file.exists()) {
            context.assets.open(assetName).use { input ->
                FileOutputStream(file).use { output ->
                    input.copyTo(output)
                }
            }
        }
        return file
    }

    companion object {
        fun formatTime(ms: Long): String {
            val seconds = (ms / 1000) % 60
            val minutes = (ms / 1000) / 60
            return String.format("%d:%02d", minutes, seconds)
        }
    }
}
